### ISSUE TITLE
How do I use postgres with Tubesync? I'm a noob when it comes to Postgres, and have just installed it.

### ISSUE THREAD

**Issue by cleverestx** [meeb/tubesync#198]
Both Tubesync and Postgres are running in containers on the same network (on a Synology, built with Docker-Compose) and both on the same bridge network. How do I get them to "talk"? Do I connect Tubesync to this database using an ENV file? Would you be willing to post an example, as I'm very new to Postgres and Tubesync. I have a ENV file set up for the password stuff, but nothing about a database is configured in any ENV yet...do add a line in the same ENV? What line? Thanks.

**Comment by meeb** [comment]
Hi, there's a full guide here already written that details how to use Postgres as a backend:

https://github.com/meeb/tubesync/blob/main/docs/other-database-backends.md

**Comment by cleverestx** [comment]
Thank you. Forgive the n00b question, but in the DATABASE_CONNECTION environmental Variable, it says for PostgreSQL to use:
postgresql://tubesync:password@localhost:5432/tubesync

Do I literally use that line, or do I need to change the first tubesync:password part to a NEW username and password, or the existing one I use with postgres? (or Tubesync's username and password?) - I'm very new to this so it's a bit confusing what it's asking for there.

Thanks again.

**Comment by meeb** [comment]
This isn't really a support forum, but I'll give you some pointers anyway. As per the linked instructions you create a database in Postgres first. Then you create a user that can access the database. When you create the user you'll set a password. It's this username and password you use in the `DATABASE_CONNECTION` line.

For example, you create a Postgres database called `tubesyncdb`, you create a user that has permissions on the database called `myuser`. You set a password with that user of `password123`. Your `DATABASE_CONNECTION` line would look like:

postgresql://myuser:password123@localhost:5432/tubesyncdb

If you're using a Postgres also in a container such as a docker-compose stack you'll also need to change `localhost` to be whatever you called the container.

You can refer to the excellent Postgres docs for more examples on creating databases and managing users.

**Comment by cleverestx** [comment]
Thank you for helping me. I think I got it connected properly to the DB. My logs for Tubesync now report something like this which is in the last part of the page you linked to test it:
"2021-04-04 22:42:17,912 [tubesync/INFO] Using database connection: django.db.backends.postgresql://tubesync:[hidden]@localhost:5432/tubesync"

But with my username and password ^ as per my Postgres database that I created also as a container in my compose file. It is running fine and reported that it's waiting for a connection in the postgres logs as well.

However, every line after that in the Tubesync logs is repeated with many of these entries:
Failed to retrieve tasks. Database unreachable
Failed to retrieve tasks. Database unreachable
Failed to retrieve tasks. Database unreachable  (over and over)

I have the ports for Postgres (and Tubesync) ALLOWED through my Synology firewall and my router, just to be comprehensive.

...as a result though, now I can no longer bring Tubesync up via the local IP or my RP address I had just previously had working for it (gives me HTTP 500 errors)...Portainer reports my Tubesync container is UNHEALTHY now as well...

I restarted it, and the logs show after the database is connected line:

`[2022-01-09 23:20:21 -0700] [460] [INFO] Starting gunicorn 20.1.0
459:C 09 Jan 2022 23:20:21.471 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
459:C 09 Jan 2022 23:20:21.471 # Redis version=6.0.16, bits=64, commit=00000000, modified=0, pid=459, just started
459:C 09 Jan 2022 23:20:21.471 # Configuration loaded
459:M 09 Jan 2022 23:20:21.474 * Running mode=standalone, port=6379.
459:M 09 Jan 2022 23:20:21.474 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
459:M 09 Jan 2022 23:20:21.474 # Server initialized
459:M 09 Jan 2022 23:20:21.474 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
459:M 09 Jan 2022 23:20:21.475 * Ready to accept connections
[2022-01-09 23:20:21 -0700] [460] [INFO] Listening at: http://127.0.0.1:8080 (460)
[2022-01-09 23:20:21 -0700] [460] [INFO] Using worker: sync
[2022-01-09 23:20:21 -0700] [490] [INFO] Booting worker with pid: 490
[2022-01-09 23:20:21 -0700] [491] [INFO] Booting worker with pid: 491
[2022-01-09 23:20:21 -0700] [492] [INFO] Booting worker with pid: 492
`
I cannot bring up Tubesync now while it is connected to Postgres. I don't know why :-(

**Comment by meeb** [comment]
If your Postgres server is running in a container then your database host name will be whatever you named the container. Like:

`postgresql://myuser:password123@postgres-container-name:5432/tubesyncdb`

I've never used Synology as a container host myself so you'd probably have to refer to the Synology docs if the above doesn't work,

**Comment by cleverestx** [comment]
Hmmm...I did rename that localhost line in the path there to my postgres container name, which I just have named to literally:  postgres

Same issue....I'm not sure what else to do. I'll keep messing with it and try to figure it out...I don't think Synology docs are going to help. They are almost useless for Docker support stuff in my experience, and I built these all with Docker-Compose files, like all my containers, not using the Synology manager itself to make it, I just have that available to toggle stuff on and off as needed basically...Everything is running on the same BRIDGE network.

I really do appreciate you trying to help. If you can think of anything else I can try, please let me know...

**Comment by meeb** [comment]
On the default bridge I don't think containers can connect to each other. They can on custom bridged networks. Try creating a custom bridged network and moving both Postgres and TubeSync to that.

https://docs.docker.com/network/bridge/#differences-between-user-defined-bridges-and-the-default-bridge

Or, using docker-compose would work fine as well.

**Comment by cleverestx** [comment]
Great idea. Both are on a new bridge network (through Synology), not via the compose, not sure how to do that in the compose. but I can look it up if that works better...I DID restart both containers too, but now in the Postgres logs there is an error, it keeps repeating:

2022-01-10 08:09:40.933 UTC [34] FATAL:  database "tubesync" does not exist.

My Tubesync logs shows:

2022-01-10 01:09:40,964 [tubesync/INFO] Using database connection: django.db.backends.postgresql://MYUSERNAME:[hidden]@postgres:5432/tubesync

Failed to retrieve tasks. Database unreachable.

Failed to retrieve tasks. Database unreachable.

**Comment by meeb** [comment]
Well, have you created the database "tubesunc" and does the MYUSERNAME user have full access to it? If you create the database then restarted the container and now the database has vanished have you set up persistent volumes for the Postgres data?

**Comment by cleverestx** [comment]
I have a persistent volume already created, but I haven't created the
database yet....but I wasn't getting that same tubesync missing error until
I changed the networks, so maybe it was created already somehow? I'm not
sure...what's the easiest way to just create the database?

**Comment by meeb** [comment]
Using a container for Postgres? Set `POSTGRES_USER` `POSTGRES_PASSWORD` and `POSTGRES_DB` ENV vars or whatever they are set to on the Postgres container. Read the Postgres container docs for examples. Then just use the same details for the connection string in TubeSync.

**Comment by cleverestx** [comment]
Thanks. That helped. I noticed my ENV file I have set up for postgres had the wrong DB name. I fixed that and the logs for postgres look good now, restarted both containers...however Tubesync logs still report (I feel like I'm so close!):

457:M 10 Jan 2022 09:42:42.204 * Ready to accept connections

2022-01-10 09:42:42,219 [tubesync/INFO] Using database connection: django.db.backends.postgresql://MYUSERNAME:[hidden]@postgres:5432/tubesync

[2022-01-10 09:42:42 -0700] [459] [INFO] Starting gunicorn 20.1.0

[2022-01-10 09:42:42 -0700] [459] [INFO] Listening at: http://127.0.0.1:8080 (459)

[2022-01-10 09:42:42 -0700] [459] [INFO] Using worker: sync

[2022-01-10 09:42:42 -0700] [491] [INFO] Booting worker with pid: 491

[2022-01-10 09:42:42 -0700] [492] [INFO] Booting worker with pid: 492

Failed to retrieve tasks. Database unreachable.

[2022-01-10 09:42:42 -0700] [493] [INFO] Booting worker with pid: 493

2022-01-10 09:42:43,346 [tubesync/INFO] Using database connection: django.db.backends.postgresql://MYUSERNAME:[hidden]@postgres:5432/tubesync

2022-01-10 09:42:43,348 [tubesync/INFO] Using database connection: django.db.backends.postgresql://MYUSERNAME:[hidden]@postgres:5432/tubesync

2022-01-10 09:42:43,373 [tubesync/INFO] Using database connection: django.db.backends.postgresql://MYUSERNAME:[hidden]@postgres:5432/tubesync

Failed to retrieve tasks. Database unreachable.

Failed to retrieve tasks. Database unreachable. (keeps repeating this)

**Comment by meeb** [comment]
Honestly I probably can't give any more pointers than I already have. It's just going to be messing about with your ENV vars and making sure the containers can connect to each other, the credentials match etc.

See if you can drop into a shell in TubeSync to debug things directly (no idea if you can do this on a Synology NAS but on normal *nix it would be `docker exec -ti tubesync bash`) and you can install telnet `apt update && apt install telnet` and see if you can connect to the postgres container directly with `telnet postgres 5432`.

You pretty much just need to get the containers be able to chat to connect to each other and make sure the credentials are correct, that's all you need to do. If you see `Database unreachable.` then your TubeSync container can't connect to whatever database server you've set in your `DATABASE_CONNECTION`.

**Comment by cleverestx** [comment]
I was able to connect via Portainer using EXEC console to the container itself...and ran apt update && apt install telnet without an issue. It's installing now..I will try the next command and see what it says. THANK YOU.

**Comment by cleverestx** [comment]
It tells me:
root@a5293f9077f8:/app# telnet postgres 5432
telnet: could not resolve postgres/5432: Name or service not known

**Comment by meeb** [comment]
Sure, that means your containers can't communicate properly as they can't resolve each others hostnames. Try a docker compose stack or tweak your custom network settings, probably the easiest way. I have no idea how to do that on Synology, however. I'm going to close this as an issue for now as it's not an issue with TubeSync and this really shouldn't turn into a general support forum. I'm sure you can work it out. You can email me if you absolutely get super stuck.

**Comment by cleverestx** [comment]
Thanks again.

**Comment by ShadowfootNZ** [comment]
This is for anyone who is having a similar problem on a mac. This has taken me a few hours to solve as a noob with installing postgres.
I have tubesync running in Docker, and I used the interactive Postgres installer from https://www.postgresql.org/download/macosx/ 

I created a database 'tubesync' and user 'tubesync'
I needed the following to set the environment variable in my docker run file for tubesync to have access to the database.
  -e DATABASE_CONNECTION='postgresql://tubesync:password@docker.for.mac.host.internal:5432/tubesync' \
